### PR TITLE
Enable loadable libraries

### PR DESCRIPTION
**Pull request by jxstelter**
This patch integrates IADK module adapter with Library manager.

### CODE REVIEW

**Review by pjdobrowolski — APPROVED** [review]

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -31,3 +31,7 @@ endif()
 if (CONFIG_MULTICORE)
 	add_subdirectory(idc)
 endif()
+
+if (CONFIG_LOADABLE_MODULES)
+	add_subdirectory(library_manager)
+endif()
```
Comment (lgirdwood) [inline src/CMakeLists.txt:37]: Needs a new line at the end.

File: zephyr/CMakeLists.txt [hunk 2]
```diff
@@ -696,6 +696,23 @@ zephyr_library_sources_ifdef(CONFIG_COMP_MODULE_ADAPTER
...
+	${SOF_SRC_PATH}/library_manager/lib_manager.c
+	${SOF_SRC_PATH}/library_manager/lib_notification.c
+)
+
+zephyr_include_directories_ifdef(CONFIG_INTEL_MODULES
+    ${SOF_SRC_PATH}/include/sof/audio/module_adapter/iadk/
+)
+
+zephyr_library_sources_ifdef(CONFIG_INTEL_MODULES
+	${SOF_AUDIO_PATH}/module_adapter/module/iadk_modules.c
```
Comment (lgirdwood) [inline zephyr/CMakeLists.txt:709]: indentation looks off here.

**Review by lgirdwood — COMMENTED** [review]
LGTM, just minor stuff.  
We should enable the Kconfig in some of the IPC4 defconfigs so that CI will build library loader now.

File: zephyr/include/rtos/string.h [hunk 3]
```diff
@@ -11,6 +11,9 @@
 #include <stddef.h>
```
Comment (dbaluta) [inline zephyr/include/rtos/string.h:11]: This patch looks good. We usually add a prefix to commit subject. 

Something like this: `zephyr: rtos`. 

Also, please imperative mood in the subject line:


e.g `zephyr: rtos: Allow memory operations to be called from cpp code`

See: https://cbea.ms/git-commit/#imperative

File: src/include/sof/lib_manager.h [hunk 4]
```diff
@@ -63,8 +63,7 @@
 #define __SOF_LIB_MANAGER_H__
 
 #include <stdint.h>
-#include <rimage/cavs/cavs_ext_manifest.h>
-#include <rimage/sof/user/manifest.h>
+#include <rimage/manifest.h>
```
Comment (dbaluta) [inline src/include/sof/lib_manager.h:67]: Same comment here. Add a prefix to the commit subject: 

e.g `library_manager: Enable loadable libraries`

File: src/audio/module_adapter/module/iadk_modules.c [hunk 5]
```diff
@@ -55,12 +55,15 @@ static int iadk_modules_init(struct processing_module *mod)
 	uint32_t module_entry_point;
 	struct module_data *md = &mod->priv;
```
Comment (dbaluta) [inline src/audio/module_adapter/module/iadk_modules.c:56]: Can you please better describe the problem here. `Latest changes in module adapter` is vague and will confuse the readers in 4 months debugging this code.

File: src/CMakeLists.txt (hunk 1 above)
Comment (jxstelter) [reply src/CMakeLists.txt]: fixed

File: zephyr/CMakeLists.txt (hunk 2 above)
Comment (jxstelter) [reply zephyr/CMakeLists.txt:709]: fixed

**Comment by jxstelter** [comment]
> LGTM, just minor stuff. We should enable the Kconfig in some of the IPC4 defconfigs so that CI will build library loader now.

Yes, I planned to add this to MTL, but MTL board is still not in upstream.

File: zephyr/include/rtos/string.h (hunk 3 above)
Comment (jxstelter) [reply zephyr/include/rtos/string.h:11]: fixed

File: src/include/sof/lib_manager.h (hunk 4 above)
Comment (jxstelter) [reply src/include/sof/lib_manager.h:67]: fixed

File: src/audio/module_adapter/module/iadk_modules.c (hunk 5 above)
Comment (jxstelter) [reply src/audio/module_adapter/module/iadk_modules.c:56]: fixed

**Review by dbaluta — APPROVED** [review]

**Comment by lgirdwood** [comment]
> > LGTM, just minor stuff. We should enable the Kconfig in some of the IPC4 defconfigs so that CI will build library loader now.
> 
> Yes, I planned to add this to MTL, but MTL board is still not in upstream.

Lets add it to TGL Kconfig on Intel side. This will give us build support, can be in another PR though.

**Review by lgirdwood — APPROVED** [review]

**Comment by jxstelter** [comment]
> > > LGTM, just minor stuff. We should enable the Kconfig in some of the IPC4 defconfigs so that CI will build library loader now.
> > 
> > 
> > Yes, I planned to add this to MTL, but MTL board is still not in upstream.
> 
> Lets add it to TGL Kconfig on Intel side. This will give us build support, can be in another PR though.

OK, will prepare PR for TGL.

**Comment by lgirdwood** [comment]
@jxstelter can you check CI, looks like some build failures with manifest CI

**Comment by marc-hb** [comment]
I just had a quick look at https://github.com/thesofproject/sof/runs/8206066990?check_suite_focus=true and all the failures are git clone failures:

```
error: RPC failed; HTTP 500 curl 22 The requested URL returned error: 500
```

Very disappointing when github is failing to git clone from github but it is what it is. Trying to re-run

**Comment by lgirdwood** [comment]
@jxstelter can you force push again to retry CI. Thanks.

**Comment by marc-hb** [comment]
Now https://sof-ci.01.org/sofpr/PR6253/build1504/build/ experienced some license server glitches... never a dull moment.

**Pull request merged**